src/libraries/System.Collections.Immutable/src/System/Collections/Frozen/Integer/DenseIntegralFrozenDictionary.cs

```diff
@@ -0,0 +1,239 @@
+            FrozenDictionary<TKey, TValue>(EqualityComparer<TKey>.Default)
+            where TKey : notnull
+            where TKeyUnderlying : IBinaryInteger<TKeyUnderlying>
+        {
+            private readonly TKey[] _keys = keys;
+            private readonly TValue[] _values = values;
+
+            private protected override TKey[] KeysCore => _keys;
+
+            private protected override TValue[] ValuesCore => _values;
+
+            private protected override int CountCore => _keys.Length;
+
+            private protected override Enumerator GetEnumeratorCore() => new Enumerator(_keys, _values);
+
+            private protected override ref readonly TValue GetValueRefOrNullRefCore(TKey key)
+            {
+                int index = int.CreateTruncating((TKeyUnderlying)(object)key);
+                TValue[] values = _values;
+                if ((uint)index < (uint)values.Length)
+                {
+                    return ref values[index];
+                }
+
+                return ref Unsafe.NullRef<TValue>();
+            }
+        }
+
+        /// <summary>Implementation used when keys are not contiguous and/or do not start at 0.</summary>
+        [DebuggerTypeProxy(typeof(DebuggerProxy<,,>))]
+        private sealed class WithOptionalValues<TKey, TKeyUnderlying, TValue>(TKey[] keys, TValue[] values, Optional<TValue>[] optionalValues, int minInclusive) :
+            FrozenDictionary<TKey, TValue>(EqualityComparer<TKey>.Default)
+            where TKey : notnull
+            where TKeyUnderlying : IBinaryInteger<TKeyUnderlying>
+        {
+            private readonly TKey[] _keys = keys;
+            private readonly TValue[] _values = values;
+            private readonly Optional<TValue>[] _optionalValues = optionalValues;
+            private readonly int _minInclusive = minInclusive;
+
+            private protected override TKey[] KeysCore => _keys;
+
+            private protected override TValue[] ValuesCore => _values;
+
+            private protected override int CountCore => _keys.Length;
+
+            private protected override Enumerator GetEnumeratorCore() => new Enumerator(_keys, _values);
+
+            private protected override ref readonly TValue GetValueRefOrNullRefCore(TKey key)
+            {
+                int keyInt32 = int.CreateTruncating((TKeyUnderlying)(object)key);
+                int minInclusive = _minInclusive;
+                if (keyInt32 >= minInclusive)
+                {
+                    int index = keyInt32 - minInclusive;
+                    Optional<TValue>[] optionalValues = _optionalValues;
+                    if ((uint)index < (uint)optionalValues.Length)
+                    {
+                        ref Optional<TValue> value = ref optionalValues[index];
+                        if (value.HasValue)
+                        {
+                            return ref value.Value;
+                        }
+                    }
+                }
```

Comment:
```suggestion
                int keyInt32 = int.CreateTruncating((TKeyUnderlying)(object)key);
                int index = keyInt32 - _minInclusive;
                Optional<TValue>[] optionalValues = _optionalValues;
                if ((uint)index < (uint)optionalValues.Length)
                {
                    ref Optional<TValue> value = ref optionalValues[index];
                    if (value.HasValue)
                    {
                        return ref value.Value;
                    }
                }
```
I guess deleting the `if (keyInt32 >= minInclusive)` should still work with the unsigned check after?